### PR TITLE
Pass metadata to gotenberg-chromium processor

### PR DESCRIPTION
**Pull request by muratbinerbay**
This will let us pass metadata to the PDF file like title, author, copyright, and creation date. 

Documentation links: 
- https://gotenberg.dev/docs/routes#metadata-chromium
- https://github.com/gotenberg/gotenberg-php/blob/a8759e59ec7d02248ac816f2f9ada7507371779e/src/Modules/ChromiumPdf.php#L173

### CODE REVIEW

**Comment by muratbinerbay** [comment]
I have read the CLA Document and I hereby sign the CLA

**Comment by muratbinerbay** [comment]
recheck

File: src/Processor/Gotenberg.php [hunk 1]
```diff
@@ -159,6 +159,10 @@ public function getPdfFromString(string $html, array $params = [], bool $returnF
             $chromium->extraHttpHeaders($params['extraHttpHeaders']);
         }
 
+        if (isset($params['metadata'])) {
+            $chromium->metadata($params['metadata']);
```
Comment (kingjia90) [inline src/Processor/Gotenberg.php:171]: Thank you for your PR!

For BC reasons, this change can't go in `1.2` as is, because this metadata was introduced in [gotenberg-php v2.2.0](https://github.com/gotenberg/gotenberg-php/releases/tag/v2.2.0)
and currently we still support v1 see https://github.com/pimcore/web-to-print-bundle/blob/f4c14bcbb4a122f6d99c93de4639b477f9d815e4/composer.json#L23

So the suggestion is to consider this as an improvement, rebase the branch to 1.x and flag this change as improvement and add a compatibility layer to avoid metadata getting called in gotenberg-php v1, as it would crash due to `Method not found` error. 
It should be using [formValue](https://github.com/gotenberg/gotenberg-php/commit/66afd63556c756e6fad5e8764ba3cec04a72998a#diff-a40bd4ab25b53d2fc6cf017d993b38d84a06232528609663bdabdd2576848705R180) on v1 instead.

**Review by kingjia90 — CHANGES_REQUESTED** [review]

File: src/Processor/Gotenberg.php (hunk 1 above)
Comment (muratbinerbay) [reply src/Processor/Gotenberg.php:171]: Hi @kingjia90, thanks for pointing that out. We missed the fact that the `metadata` method is not available on v1.
I updated the code to support both versions.

File: src/Processor/Gotenberg.php [hunk 2]
```diff
@@ -159,6 +159,12 @@ public function getPdfFromString(string $html, array $params = [], bool $returnF
             $chromium->extraHttpHeaders($params['extraHttpHeaders']);
         }
 
+        if (isset($params['metadata']) && method_exists($chromium, 'metadata')) {
+            $chromium->metadata($params['metadata']);
+        } elseif (isset($params['metadata'])) {
+            $chromium->formValue('metadata', json_encode($params['metadata']));
+        }
```
Comment (blankse) [inline src/Processor/Gotenberg.php]: I think it is enough to check the existence of metadata once.
```suggestion
        if (isset($params['metadata'])) {
            if (method_exists($chromium, 'metadata')) {
                $chromium->metadata($params['metadata']);
            } else {
                $chromium->formValue('metadata', json_encode($params['metadata']));
            }
        }
```

File: src/Processor/Gotenberg.php [hunk 3]
```diff
@@ -159,6 +159,14 @@ public function getPdfFromString(string $html, array $params = [], bool $returnF
             $chromium->extraHttpHeaders($params['extraHttpHeaders']);
         }
 
+        if (isset($params['metadata'])) {
+            if (method_exists($chromium, 'metadata')) {
+                $chromium->metadata($params['metadata']);
+            } else {
+                $chromium->formValue('metadata', json_encode($params['metadata']));
+            }
```
Comment (kingjia90) [inline src/Processor/Gotenberg.php]:
```suggestion
          // TODO: Replace with metadata() once dropping support for gotenberg-php < 2.2.0
          $chromium->formValue('metadata', json_encode($params['metadata']));
            
```
Thank you both for suggesting the changes.

It seems that this way might be slightly better, "less prettier" than using the direct method, but at least is avoiding `method_exists`.

Comment (muratbinerbay) [reply src/Processor/Gotenberg.php]: I thought of that but wanted to avoid the `TODO` :) 
Looks much cleaner I guess! Thanks

**Comment by muratbinerbay** [comment]
@kingjia90 on v2.2.0 the `formValue` has protected access. I reverted last 2 commits to what @blankse suggested.

https://github.com/gotenberg/gotenberg-php/blob/main/src/MultipartFormDataModule.php#L92

**Comment by blankse** [comment]
@muratbinerbay @kingjia90 The formValue is only in 1.1.8 public.
See: 
https://github.com/gotenberg/gotenberg-php/commit/33298eb81eb8e63fb0aaca71607f98b553a3bb50
https://github.com/gotenberg/gotenberg-php/commit/993384d30dadc6cb74b8e2daa1355756f9fbf331
So we need to bump the version in composer, right?
`"gotenberg/gotenberg-php": "^1.1.8 || ^2.2",`

**Comment by kingjia90** [comment]
So many plot twists 😄 Thank you for the changes. 
It seems that PHPStan is triggering some false positive i guess, do we need to mute these?

**Comment by blankse** [comment]
@kingjia90 Yes PhpStan doesn't want the formValue method because it is protected in 2.2. But it isn't reachable there.

File: src/Processor/Gotenberg.php [hunk 4]
```diff
@@ -159,6 +159,15 @@ public function getPdfFromString(string $html, array $params = [], bool $returnF
             $chromium->extraHttpHeaders($params['extraHttpHeaders']);
         }
 
+        if (isset($params['metadata'])) {
+            if (method_exists($chromium, 'metadata')) {
+                $chromium->metadata($params['metadata']);
+            } else {
+                /* @phpstan-ignore-next-line */
+                $chromium->formValue('metadata', json_encode($params['metadata']));
```
Comment (blankse) [inline src/Processor/Gotenberg.php]: Can you try this?
```suggestion
            } elseif (is_callable([$chromium, 'formValue'])) {
                $chromium->formValue('metadata', json_encode($params['metadata']));
```

**Comment by muratbinerbay** [comment]
@blankse @kingjia90 do we want to try something like [ReflectionMethod](https://www.php.net/manual/en/class.reflectionmethod.php). Seems rather excessive to me but I'm willing to try it :)

**Comment by blankse** [comment]
Doesn't work :(

I think we need the same construct like in pimcore/pimcore. We have there a own phpstan config for lowest with `reportUnmatchedIgnoredErrors: false`:
https://github.com/pimcore/pimcore/blob/11.x/phpstan-lowest.neon

So we can create a phpstan baseline with the highest dependencies.

**Comment by blankse** [comment]
Of course, we could also process the metadata with version 2.2 only, if that is an alternative :)

**Comment by muratbinerbay** [comment]
> Of course, we could also process the metadata with version 2.2 only, if that is an alternative :)

would also be acceptable in my opinion, @kingjia90 what do you think?

**Comment by kingjia90** [comment]
That would be also fine to me, no any objection, but it seems that then it would need a doc or changelog entry (as [example](https://github.com/pimcore/customer-data-framework/blob/4.x/CHANGELOG.md)) to better explain that "difference" (that metadata is available only if you use gotenberg-php 2.2+ )

maybe let's have a final confirmation by @fashxp first on how to proceed

**Comment by muratbinerbay** [comment]
Hi @kingjia90 @fashxp, just wanted to ping you to check if I can go ahead and make the metadata available only for gotenberg-php:2.2+ ?
Thanks!

**Comment by kingjia90** [comment]
Thank you for your patience @muratbinerbay !

I think it would be fine to support metadata only for 2.2+ and we can proceed with that, we also have a similar situation with the [screenshot](https://github.com/pimcore/pimcore/pull/16608) feature in 2.0 and it's ["diverging"](https://github.com/pimcore/pimcore/pull/17023) more and more in what one version can do and what not, so at some point we just need to properly document the difference and recommend to use the latest when possible.

File: composer.json [hunk 5]
```diff
@@ -20,7 +20,7 @@
   "require-dev": {
     "phpstan/phpstan": "^1.10.5",
     "phpstan/phpstan-symfony": "^1.2.20",
-    "gotenberg/gotenberg-php": "^1.0 || ^2.0",
+    "gotenberg/gotenberg-php": "^1.1.8 || ^2.2",
```
Comment (muratbinerbay) [inline composer.json:23]: @kingjia90 should I revert this change in this case? I guess versions `2.0` and `2.1` are also acceptable, but the metadata won't work.

Comment (kingjia90) [reply composer.json:23]: Thank you, mhhh, i personally would keep it, as long as it's documented in the minor release changelog, we should encourage to use a more recent dependency, and it would be simpler and easier for devs to keep in mind that the `metadata()` feature is just NOT available to `gotenberg-php` `1.x`

Comment (muratbinerbay) [reply composer.json:23]: Makes sense 👍, resolving
